src/llvm-propagate-addrspaces.cpp

```diff
@@ -0,0 +1,205 @@
+    PropagateJuliaAddrspaces() : FunctionPass(ID) {};
+
+public:
+    bool runOnFunction(Function &F) override;
+    Value *LiftPointer(Value *V, Type *LocTy = nullptr, Instruction *InsertPt=nullptr);
+    void visitStoreInst(StoreInst &SI);
+    void visitLoadInst(LoadInst &LI);
+    void visitMemSetInst(MemSetInst &MI);
+    void visitMemTransferInst(MemTransferInst &MTI);
+};
+
+bool PropagateJuliaAddrspaces::runOnFunction(Function &F) {
+    visit(F);
+    for (auto it : ToInsert)
+        it.first->insertBefore(it.second);
+    for (Instruction *I : ToDelete)
+        I->eraseFromParent();
+    ToInsert.clear();
+    ToDelete.clear();
+    LiftingMap.clear();
+    Visited.clear();
+    return true;
+}
+
+static unsigned getValueAddrSpace(Value *V) {
+    return cast<PointerType>(V->getType())->getAddressSpace();
+}
+
+static bool isSpecialAS(unsigned AS) {
+    return AddressSpace::FirstSpecial <= AS && AS <= AddressSpace::LastSpecial;
+}
+
+Value *PropagateJuliaAddrspaces::LiftPointer(Value *V, Type *LocTy, Instruction *InsertPt) {
+    SmallVector<Value *, 4> Stack;
+    Value *CurrentV = V;
+    // Follow pointer casts back, see if we're based on a pointer in
+    // an untracked address space, in which case we're allowed to drop
+    // intermediate addrspace casts.
+    while (true) {
+        Stack.push_back(CurrentV);
+        if (isa<BitCastInst>(CurrentV))
+            CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
+        else if (isa<AddrSpaceCastInst>(CurrentV)) {
+            CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
+            if (!isSpecialAS(getValueAddrSpace(CurrentV)))
+                break;
+        }
+        else if (isa<GetElementPtrInst>(CurrentV)) {
+            if (LiftingMap.count(CurrentV)) {
+                CurrentV = LiftingMap[CurrentV];
+                break;
+            } else if (Visited.count(CurrentV)) {
+                return nullptr;
+            }
+            Visited.insert(CurrentV);
+            CurrentV = cast<GetElementPtrInst>(CurrentV)->getOperand(0);
+        } else
+            break;
+    }
+    if (!CurrentV->getType()->isPointerTy())
+        return nullptr;
+    if (isSpecialAS(getValueAddrSpace(CurrentV)))
+        return nullptr;
+    // Ok, we're allowed to change the address space of this load, go back and
+    // reconstitute any GEPs in the new address space.
+    for (Value *V : llvm::reverse(Stack)) {
+        GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(V);
+        if (!GEP)
+            continue;
+        if (LiftingMap.count(GEP)) {
+            CurrentV = LiftingMap[GEP];
+            continue;
+        }
+        GetElementPtrInst *NewGEP = cast<GetElementPtrInst>(GEP->clone());
+        ToInsert.push_back(std::make_pair(NewGEP, GEP));
+        Type *GEPTy = GEP->getSourceElementType();
+        Type *NewRetTy = cast<PointerType>(GEP->getType())->getElementType()->getPointerTo(getValueAddrSpace(CurrentV));
```

Comment:
If you're putting an addrspace on this, there should be one on the BitCast below too, otherwise the BitCast inst would be attempting to change AddrSpace also.